Packs/Phishing/Scripts/PhishingDedupPreprocessingRule/PhishingDedupPreprocessingRule.py

```diff
@@ -0,0 +1,209 @@
+    incidents_df[MERGED_TEXT_FIELD] = incidents_df.apply(lambda x: preprocess_text_fields(x), axis=1)
+    incidents_df = incidents_df[incidents_df[MERGED_TEXT_FIELD].str.len() >= MIN_TEXT_LENGTH]
+    incidents_df.reset_index(inplace=True)
+    if FROM_FIELD in incidents_df:
+        incidents_df[FROM_FIELD] = incidents_df[FROM_FIELD].fillna(value='')
+    else:
+        incidents_df[FROM_FIELD] = ''
+    incidents_df[FROM_FIELD] = incidents_df[FROM_FIELD].apply(lambda x: x.strip())
+    incidents_df[FROM_DOMAIN_FIELD] = incidents_df[FROM_FIELD].apply(lambda address: extract_domain(address))
+    return incidents_df
+
+
+def filter_out_new_incident(existing_incidents_df, new_incident):
+    same_id_mask = existing_incidents_df['id'] == new_incident['id']
+    existing_incidents_df = existing_incidents_df[~same_id_mask]
+    return existing_incidents_df
+
+
+def vectorize(text, vectorizer):
+    return vectorizer.transform([text]).toarray()[0]
+
+
+def cosine_sim(a, b):
+    return dot(a, b) / (norm(a) * norm(b))
+
+
+def find_duplicate_incidents(new_incident, existing_incidents_df):
+    global MERGED_TEXT_FIELD, FROM_POLICY
+    new_incident_text = new_incident[MERGED_TEXT_FIELD]
+    text = [new_incident_text] + existing_incidents_df[MERGED_TEXT_FIELD].tolist()
+    vectorizer = TfidfVectorizer(token_pattern=r"(?u)\b\w\w+\b|!|\?|\"|\'").fit(text)
+    new_incident_vector = vectorize(new_incident_text, vectorizer)
+    existing_incidents_df['vector'] = existing_incidents_df[MERGED_TEXT_FIELD].apply(lambda x: vectorize(x, vectorizer))
+    existing_incidents_df['similarity'] = existing_incidents_df['vector'].apply(
+        lambda x: cosine_sim(x, new_incident_vector))
+    if FROM_POLICY == FROM_POLICY_DOMAIN:
+        mask = (existing_incidents_df[FROM_DOMAIN_FIELD] != '') &\
+               (existing_incidents_df[FROM_DOMAIN_FIELD] == new_incident[FROM_DOMAIN_FIELD])
+        existing_incidents_df = existing_incidents_df[mask]
+    elif FROM_POLICY == FROM_POLICY_EXACT:
+        mask = (existing_incidents_df[FROM_FIELD] != '') & \
+               (existing_incidents_df[FROM_FIELD] == new_incident[FROM_FIELD])
+        existing_incidents_df = existing_incidents_df[mask]
+    existing_incidents_df.sort_values(by='similarity', ascending=False, inplace=True)
+    if len(existing_incidents_df) > 0:
+        return existing_incidents_df.iloc[0], existing_incidents_df.iloc[0]['similarity']
+    else:
+        return None, None
+
+
+def close_new_incident_and_link_to_existing(new_incident, existing_incident, similarity):
+    entries = []
+    # dumpable_new_incident = {k: v for k, v in new_incident.items() if not isinstance(v, (np.floating, np.integer))}
+    entries.append({'Contents': "Duplicate incident: " + new_incident['name']})
+    entries.append({"Type": entryTypes['note'],
+                    "ContentsFormat": "json",
+                    "Contents": json.dumps(new_incident)
```

Comment:
Do you think we need all incident data as JSON? not sure it's the best way... also not sure you need to dump. you should consider create human readable for that, convert the new_incident to table with tableToMarkdown

Reply:
filtered out only a few fields